### ISSUE TITLE
403 error from galaxy api

### ISSUE THREAD

**Issue by moskalenko** [galaxyproject/galaxy#9916]
Galaxy version: 20_01
Galaxy can be accessed after clearing all browser cookies.
Ostensibly, after visiting certain university webpages galaxy api authentication stops working resulting in 403 errors when getting the history and such. The only difference I can discern is in which cookies are passed in the header. I was not able to home in on a particular offending cookie. While deleting non-galaxy cookies one by one and reloading the galaxy main page authentication starts working at some point without regard to the order of cookie deletion. Here is an example of a raw request header when the api 403 error  is encountered. I've redacted potentially sensitive information from non-galaxy cookies with xxxxxxxxx. The only cookies required to access galaxy are galaxysession and auth-tkt (single signon auth cookie). One of the other cookies seems to trip galaxy api and cause a 403 error. This is a production galaxy instance issue only encountered after the upgrade to 20_01.

```
Host: galaxy.rc.xxxxxxx
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:77.0) Gecko/20100101 Firefox/77.0
Accept: application/json, text/plain, */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
DNT: 1
Connection: keep-alive
Referer: https://xxxxxxxxxxxx
Cookie: galaxysession=55add6616e377be61f5499f63f0ba60813fdbb3ff4e6c3fcc0fd6ccf9f78413b7e9445ae08da01e8; experimentation_subject_id=ImVjYTQ0ODc5LWE2OTEtNDI5Mi04YTNmLTQzZjZlOTA2NWFjMSI%3D--909ae45a162008b836a433ec3e8c69474b64fdfc; PS_DEVICEFEATURES=width:2560 height:1440 pixelratio:1 touch:0 geolocation:1 websockets:1 webworkers:1 datepicker:1 dtpicker:0 timepicker:1 dnd:1 sessionstorage:1 localstorage:1 history:1 canvas:1 svg:1 postmessage:1 hc:0 maf:0; psping=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx; prod-ps-web1-33200-PORTAL-PSJSESSIONID=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx; PS_LOGINLIST=xxxxxxxx; PS_TOKENEXPIRE=22_Jun_2020_19:45:07_GMT; lcsrftoken=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx; PS_TokenSite=xxxxxxxxxxxxxxxxxxxxxxxxxxx; SignOnDefault=; HPTabName=DEFAULT; HPTabNameRemote=; LastActiveTab=DEFAULT; BIGipServer~ESI~xxxxxxxx-pool=xxxxxxxxxxxxxxx; BIGipServer~ESI~xxxxxxxxxxxxxx=xxxxxxxxxxxxxx; BIGipServer~ESI~xxxxxxxxxxxxxx=2883xxxxxxxxxxxxx; ExpirePage=xxxxxxxxxxxx; PS_TOKEN=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx=; ps_theme=node:EMPL portal:EMPLOYEE theme_id:UF_DEFAULT_THEME_FLUID css:UF_BRAND_CLASSIC_TEMPLTE_FLUID css_f:UF_BRAND_FLUID_TEMPLATE accessibility:N macroset:UF_DEFAULT_MACROSET_855 formfactor:3 piamode:2; xxxxxxxxxxxxxx-PORTAL-PSJSESSIONID=xxxxxxxxxxxxxxxxxxxxxxxxxxxx; prod-xxxxxxxxxxx-PORTAL-PSJSESSIONID=xxxxxxxxxxxxxxxxxxxxxxxxxx; https%3a%2f%2fxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx%3Dremoteunifieddashboard%7C%7C; psback=%22%22url%22%3A%22https%3A%2F%2Fxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx%22%20%22origin%22%3A%22PIA%22%20%22layout%22%3A%221%22%22; xxxxxxxxxxxxxxxxxxxx-33700-PORTAL-PSJSESSIONID=xxxxxxxxxxxxxxxxxxxxxxxx; auth_tkt=xxxxxxxxxxxxxxxxxxxxxxxx28%3D
Cache-Control: max-age=0
```

**Comment by moskalenko** [comment]
Here's one walk down the edit-header-resend lane I made. From the above list of cookies I whittled the request down to 

403 response for the following cookies in the  header:
Cookie: galaxysession=55add66xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxeedaaea5c508; auth_tkt=ZjM0MjViMxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxVua28%3D; PS_TOKEN=oQAAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx3UzLs6+wB/6OIuxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxzfDhiCpo=; ps_theme=node:EMPL portal:EMPLOYEE theme_id:UF_DEFAULT_THEME_FLUID css:UF_BRAND_CLASSIC_TEMPLTE_FLUID css_f:UF_BRAND_FLUID_TEMPLATE accessibility:N macroset:UF_DEFAULT_MACROSET_855 formfactor:3 piamode:2; ps_theme=node:EMPL portal:EMPLOYEE theme_id:UF_DEFAULT_THEME_FLUID css:UF_BRAND_CLASSIC_TEMPLTE_FLUID css_f:UF_BRAND_FLUID_TEMPLATE accessibility:N macroset:UF_DEFAULT_MACROSET_855 formfactor:3 piamode:2;

200 response without 
ps_theme=node:EMPL portal:EMPLOYEE theme_id:UF_DEFAULT_THEME_FLUID css:UF_BRAND_CLASSIC_TEMPLTE_FLUID css_f:UF_BRAND_FLUID_TEMPLATE accessibility:N macroset:UF_DEFAULT_MACROSET_855 formfactor:3 piamode:2;

with the following cookies in the header:
Cookie: galaxysession=55add66xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxeedaaea5c508; auth_tkt=ZjM0MjViMxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxVua28%3D; PS_TOKEN=oQAAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx3UzLs6+wB/6OIuxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxzfDhiCpo=; 

So, the minimal cookie set in the request header to trigger a 403 from galaxy API in this 'cookie walk' is 

Cookie: galaxysession=55add6xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx5c508; auth_tkt=ZjM0MjxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxVua28%3D; ps_theme=node:EMPL portal:EMPLOYEE theme_id:UF_DEFAULT_THEME_FLUID css:UF_BRAND_CLASSIC_TEMPLTE_FLUID css_f:UF_BRAND_FLUID_TEMPLATE accessibility:N macroset:UF_DEFAULT_MACROSET_855 formfactor:3 piamode:2;

So, in this example galaxy api auth breaks on having the following cookie in the header.

portal:EMPLOYEE theme_id:UF_DEFAULT_THEME_FLUID css:UF_BRAND_CLASSIC_TEMPLTE_FLUID css_f:UF_BRAND_FLUID_TEMPLATE accessibility:N macroset:UF_DEFAULT_MACROSET_855 formfactor:3 piamode:2;

**Comment by dannon** [comment]
Thanks for all the details; taking a look now.

**Comment by moskalenko** [comment]
Another cookie type that trips galaxy auth api is non-url-encoded path. E.g. a 403 error is returned if the following cookie is re-added to a minimal clean header:
PS_LOGINLIST=https://f.q.d.n/ps https://f.q.d.n/ps;

but if it's encoded into 

PS_LOGINLIST=https%3A%2F%2Ff.q.d.n%2Fps%20https%3A%2F%2Ff.q.d.n%2Fps;

a 200 response is returned

**Comment by dannon** [comment]
Yep, to close the loop on the details here (for myself and others), we tracked this down to what seem to be likely malformed cookies per https://tools.ietf.org/html/rfc6265#section-4.1.1.  That this used to work (19.05) and now doesn't (20.01) makes me think we updated a web stack dependency that is now more strict, or otherwise tweaked cookie handling (perhaps for py3 compatibility?) and have introduced this regression (technically, even though I think Galaxy is following the spec).

With this, I can trace through with a malformatted cookie and come up with a better fix.

**Comment by dannon** [comment]
Relevant python issue that @moskalenko found:  
https://bugs.python.org/issue31456

tldr; python stdlib SimpleCookie throws away the entire cookie without raising an exception if there is any whitespace.  We can either accept this, or add a patch sanitizing the galaxysession cookie.

**Comment by moskalenko** [comment]
Patched base py with the discussed fix improved by @dannon. As far as  I can see this resolves the issue. I don't see any side effects.


diff --git a/lib/galaxy/web/framework/base.py b/lib/galaxy/web/framework/base.py
index edf1b22b97..9458f0911c 100644
--- a/lib/galaxy/web/framework/base.py
+++ b/lib/galaxy/web/framework/base.py
@@ -388,7 +388,7 @@ class Request(webob.Request):
 
     @lazy_property
     def cookies(self):
-        return get_cookies(self.environ)
+        return get_cookies({"HTTP_COOKIE": "; ".join([x.strip() for x in self.environ.get("HTTP_COOKIE").split('; ') if x.startswith('galaxy')])})
 
     @lazy_property
     def base(self):

**Comment by dannon** [comment]
This was fixed in https://github.com/galaxyproject/galaxy/pull/9921
